x-pack/plugin/wildcard/src/main/java/org/elasticsearch/xpack/wildcard/mapper/WildcardFieldMapper.java

```diff
@@ -0,0 +1,541 @@
+                        // *foo by indexing smaller ngrams at the end of a string
+                        TermQuery tq = new TermQuery(new Term(name(), fragment));
+                        bqBuilder.add(new BooleanClause(tq, Occur.MUST));
+                    }
+                }
+            }
+            
+            BooleanQuery approximation = bqBuilder.build();
+            if (patternStructure.isMatchAll()) {
+                return new MatchAllDocsQuery();
+            } 
+            if (approximation.clauses().size() > 1 || patternStructure.needsVerification()) {
+                BooleanQuery.Builder verifyingBuilder = new BooleanQuery.Builder();
+                verifyingBuilder.add(new BooleanClause(approximation, Occur.MUST));
+                Automaton automaton = WildcardQuery.toAutomaton(new Term(name(), wildcardPattern));
+                verifyingBuilder.add(new BooleanClause(new AutomatonQueryOnBinaryDv(name(), wildcardPattern, automaton), Occur.MUST));
+                return verifyingBuilder.build();
+            }
+            return approximation;
+        }                
+
+        int numChars() {
+            return numChars;
+        }
+
+        void setNumChars(int numChars) {
+            checkIfFrozen();
+            this.numChars = numChars;
+        }
+        
+        @Override
+        public void checkCompatibility(MappedFieldType fieldType, List<String> conflicts) {
+            super.checkCompatibility(fieldType, conflicts);
+            WildcardFieldType other = (WildcardFieldType)fieldType;
+            // prevent user from changing num_chars
+            if (numChars() != other.numChars()) {
+                conflicts.add("mapper [" + name() + "] has different [num_chars]");
+            }
+        }        
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            return new DocValuesFieldExistsQuery(name());
+        }
+
+        @Override
+        public Query termQuery(Object value, QueryShardContext context) {
+            return wildcardQuery(BytesRefs.toString(value), MultiTermQuery.CONSTANT_SCORE_REWRITE, context);
+        }
+
+        @Override
+        public Query termsQuery(List<?> values, QueryShardContext context) {
+            BooleanQuery.Builder bq = new BooleanQuery.Builder();
+            for (Object value : values) {
+                bq.add(termQuery(value, context), Occur.SHOULD);
+            }
+            return new ConstantScoreQuery(bq.build());
+        }     
+                
+        @Override
+        public IndexFieldData.Builder fielddataBuilder(String fullyQualifiedIndexName) {
+            failIfNoDocValues();
+            return new DocValuesIndexFieldData.BinaryBuilder();
+        }
+        
+    }
+
+    private int ignoreAbove;
+    private int numChars;
```

Comment:
let's not store `numChars` here, I'd rather like to call `fieldType().numChars()` whenever we need it from the mapper, it avoids having to think about keeping them in sync